src/codegate/ca/codegate_ca.py

```diff
@@ -1,18 +1,85 @@
-import datetime
 import os
 import ssl
+import threading
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone
 from typing import Dict, Optional, Tuple
 
 import structlog
 from cryptography import x509
+from cryptography.exceptions import InvalidSignature
+from cryptography.hazmat.backends import default_backend
 from cryptography.hazmat.primitives import hashes, serialization
-from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.hazmat.primitives.asymmetric import rsa, padding
 from cryptography.x509.oid import ExtendedKeyUsageOID, NameOID
 
 from codegate.config import Config
 
 logger = structlog.get_logger("codegate")
 
+# Add a buffer to renew certificates slightly before expiry.
+TLS_GRACE_PERIOD_DAYS = 2
+
+
+@dataclass
+class CachedCertificate:
+    """Hold certificate data with metadata"""
+
+    cert_path: str
+    key_path: str
+    creation_time: datetime
+
+
+class TLSCertDomainManager:
+    """
+    This class manages SSL contexts for domain certificates with SNI
+    """
+
+    def __init__(self, ca_provider: "CertificateAuthority"):
+        self._ca = ca_provider
+        # Use strong references for caching
+        self._cert_cache: Dict[str, CachedCertificate] = {}
+        self._context_cache: Dict[str, ssl.SSLContext] = {}
```

Comment:
This is not for this PR, but would we ever want to do some cache cleanup like least-recently used or are the certificates small enough that we don't care?

Reply:
I think so, I will follow up, just want to think it over a bit more